contracts/src/registry/L1ETHRegistry.sol

```diff
@@ -0,0 +1,244 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+
+import {ERC1155Singleton} from "./ERC1155Singleton.sol";
+import {IERC1155Singleton} from "./IERC1155Singleton.sol";
+import {IRegistry} from "./IRegistry.sol";
+import {IRegistryDatastore} from "./IRegistryDatastore.sol";
+import {BaseRegistry} from "./BaseRegistry.sol";
+import {PermissionedRegistry} from "./PermissionedRegistry.sol";
+
+/**
+ * @title L1ETHRegistry
+ * @dev L1 contract for .eth that holds ejected .eth names only.
+ * Unlike the L2 ETHRegistry, this registry does not handle new registrations directly,
+ * but receives names that have been ejected from L2.
+ */
+contract L1ETHRegistry is PermissionedRegistry, AccessControl {
+    bytes32 public constant EJECTION_CONTROLLER_ROLE = keccak256("EJECTION_CONTROLLER_ROLE");
+    bytes32 public constant RENEWAL_CONTROLLER_ROLE = keccak256("RENEWAL_CONTROLLER_ROLE");
+
+    error NameExpired(uint256 tokenId);
+    error CannotReduceExpiration(uint64 oldExpiration, uint64 newExpiration);
+
+    event NameEjected(uint256 indexed tokenId, address owner, uint64 expires);
+    event NameRenewed(uint256 indexed tokenId, uint64 newExpiration, address renewedBy);
+    event NameRelinquished(uint256 indexed tokenId, address relinquishedBy);
+    event NameMigratedToL2(uint256 indexed tokenId, address sendTo);
+    event FallbackResolverSet(address resolver);
+    event TokenObserverSet(uint256 indexed tokenId, address observer);
+
+    // Address of the fallback resolver for names not found in this registry
+    address public fallbackResolver;
+
+    // Map to track token observers for notification of renewal/ejection events
+    mapping(uint256 => address) public tokenObservers;
+
+    constructor(IRegistryDatastore _datastore) PermissionedRegistry(_datastore) {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function uri(uint256 /*tokenId*/ ) public pure override returns (string memory) {
+        return "";
+    }
+
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
+    {
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(oldFlags);
+        if (expires < block.timestamp) {
+            return address(0);
+        }
+        return super.ownerOf(tokenId);
+    }
+
+    /**
+     * @dev Receive an ejected name from L2.
+     * @param label The label of the name
+     * @param owner The owner of the name
+     * @param registry The registry to use for the name
+     * @param flags Additional flags for the name
+     * @param expires Expiration timestamp
+     * @return tokenId The token ID of the ejected name
+     */
+    function ejectFromL2(string calldata label, address owner, IRegistry registry, uint96 flags, uint64 expires)
```

Comment:
I think we should probably be using the label hash here, not the plaintext label.